samza-core/src/main/java/org/apache/samza/container/RunLoop.java

```diff
@@ -805,6 +877,52 @@ private boolean checkEndOfStream() {
       return processingSspSet.isEmpty();
     }
 
+    private boolean shouldDrain() {
+      if (endOfStream) {
+        return false;
+      }
+
+      if (!pendingEnvelopeQueue.isEmpty()) {
+        PendingEnvelope pendingEnvelope = pendingEnvelopeQueue.peek();
+        IncomingMessageEnvelope envelope = pendingEnvelope.envelope;
+
+        if (envelope.isDrain()) {
+          final DrainMessage message = (DrainMessage) envelope.getMessage();
+          if (!message.getDeploymentId().equals(deploymentId)) {
+            // Removing the drain message from the pending queue as it doesn't match with the current deploymentId
+            // Removing it will ensure that it is not picked up by process()
+            pendingEnvelopeQueue.remove();
+          } else {
+            if (elasticityFactor <= 1) {
+              SystemStreamPartition ssp = envelope.getSystemStreamPartition();
+              processingSspSetToDrain.remove(ssp);
+            } else {
+              // if envelope contains drain message,
```

Comment:
Maybe a little bit more explaination here about the log to handle elasticity. Seems this feature is making things much complex.

Reply:
Added extra explanation.